### PR TITLE
[mlir][gpu] Align reduction operations with vector combining kinds

### PR DESCRIPTION
**Pull request by kuhar**
The motivation for this change is explained in
https://github.com/llvm/llvm-project/issues/72354.

Before this change, we could not tell between signed/unsigned minimum/maximum and NaN treatment for floating point values.

The mapping of old reduction operations to the new ones is as follows:
*  `min` --> `minsi` for ints, `minf` for floats
*  `max` --> `maxsi` for ints, `maxf` for floats

New reduction kinds not represented in the old enum: `minui`, `maxui`, `minimumf`, `maximumf`.

As a next step, I would like to have a common definition of combining kinds used by the `vector` and `gpu` dialects. Separately, the GPU to SPIR-V lowering does not yet properly handle zero and NaN values -- the behavior of floating point min/max group reductions is not specified by the SPIR-V spec.

Issue: https://github.com/llvm/llvm-project/issues/72354

### CODE REVIEW

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp
Comment (antiagainst) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:506]: The SPIR-V spec is just providing the syntax and base semantics for those ops. For detailed numeric precision requirements, we need to look at the target client API's spec. For Vulkan there are additional spec on ["Precision and Operation of SPIR-V Instructions"](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/chap52.html#spirvenv-precision-operation):

> By default, the implementation may perform optimizations on half, single, or double-precision floating-point instructions that ignore sign of a zero, or assume that arguments and results are not NaNs or infinities. If the entry point is declared with the SignedZeroInfNanPreserve Execution Mode, then NaNs, infinities, and the sign of zero must not be ignored.

I think in general we are a bit handwavy (due to the fuzziness of ML compute) regarding this part in MLIR (as we are continuing fixing various layers about these numerics; this patch being one of them) and so in SPIR-V. I won't block this pull request due to that given this makes the situation better. But could you create an issue to track the SPIR-V side to better handle floating point numerics (specifcially we need to audit how we handle fast math, emit `SignedZeroInfNanPreserve` if requested, etc.)? We aren't doing that right now.

**Review by antiagainst — APPROVED** [review]
LGTM! But please wait for other stakeholders to take a look before merging. :)

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp
Comment (kuhar) [reply mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp]: Sure, I filed https://github.com/llvm/llvm-project/issues/73459 and pointed to it in the TODO comment.

Comment (antiagainst) [reply mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp]: Thanks @kuhar!!

File: mlir/lib/Dialect/GPU/IR/GPUDialect.cpp
Comment (Hardcode84) [inline mlir/lib/Dialect/GPU/IR/GPUDialect.cpp:505]: Can we add explicit checks `isa<IntegerType, FloatType>` for `add` and `mul` reductions too? Op definition in .td takes `AnyType` but I think it is an oversight.

Comment (kuhar) [reply mlir/lib/Dialect/GPU/IR/GPUDialect.cpp:505]: I added extra type constraints in the op definition

**Review by Hardcode84 — APPROVED** [review]

**Pull request merged**
